### DISCUSSION TITLE
Accumulate vs Reduce vs Foldl

### DISCUSSION THREAD

**Discussion by leshabirukov** [Ilykuleshov/pytorch-toolz#1] [General]
Greetings!
Well, I looked at RNN from tutorial (https://github.com/Ilykuleshov/pytorch-toolz/blob/main/examples/RNN.ipynb), currently it implemented like:
```
  rnn = Accumulate(
      lstm_swap_args,
      initial=(torch.zeros(8),torch.zeros(8))
  )
```
Early there was `Reduce` in place of `Accumulate`, and I hesitate, is it really better? `Accumulate` more like assumes associativity, commutativity etc. My favorite is actually `Foldl` as in `Haskell` because even `Foldr` will not go. Off cource, it is just opinion.

-------------------
EDIT:
Oh, I see, it descends from standart itertools semantics! That's closes the question.

**Comment by Ilykuleshov** [comment]
I agree, these Haskell functions would definitely come in useful, but I am trying to base this lib on standard python, so as to eliminate possible clutter of non-standard functions

**Reply by leshabirukov** [reply]
Ok, but look at `functools.reduce` signature: `def reduce(function, iterable, initializer=None):` The last parameter name is `initializer`, not `initial` as in `itertools.accumulate`.  
And one more question: what `, function: Callable, **/**) ->` is for, some arbitrary future extension, or you have more definite plans for it?

**Reply by Ilykuleshov** [reply]
Interesting, in my version of python (3.9.12) the definition is `def reduce(function, sequence, initial=_initial_missing):`. Anyway, its a positional-only argument, so it shouldn't matter
As for the slash, its the [syntax to indicate the end of positional-only arguments](https://stackoverflow.com/questions/24735311/what-does-the-slash-mean-when-help-is-listing-method-signatures)
